tools/c7n_azure/c7n_azure/functionapp_utils.py

```diff
@@ -33,6 +35,102 @@ def generate_machine_decryption_key():
+
+    def deploy_infrastructure(self, parameters):
+        # Check if RG exists
+        session = self.local_session
+        rg_client = session.client('azure.mgmt.resource.ResourceManagementClient')
+        if not rg_client.resource_groups.check_existence(parameters.group_name):
+            self.log.info("Deploying resource group %s" % (parameters.group_name))
+            rg_client.resource_groups.create_or_update(parameters.group_name,
+                                                       {'location': parameters.location})
+        else:
+            self.log.info("Found resource group %s" % (parameters.group_name))
+
+        # Storage account create function is async, wait for completion after
+        # other resources provisioned.
+        sm_client = session.client('azure.mgmt.storage.StorageManagementClient')
+        accounts = sm_client.storage_accounts.list_by_resource_group(parameters.group_name)
+        found = parameters.storage_name in [a.name for a in accounts]
+        account = None
+        if not found:
+            self.log.info("Deploying storage account %s in %s resource group" %
+                          (parameters.storage_name, parameters.group_name))
+            params = {'sku': {'name': 'Standard_LRS'},
+                      'kind': 'Storage',
+                      'location': parameters.location}
+            account = sm_client.storage_accounts.create(parameters.group_name,
+                                                        parameters.storage_name,
+                                                        params)
+        else:
+            self.log.info("Found storage account %s in %s resource group" %
+                          (parameters.storage_name, parameters.group_name))
+
+        # Deploy app insights if needed
+        ai_client = \
+            session.client(
+                'azure.mgmt.applicationinsights.ApplicationInsightsManagementClient')
+        try:
+            ai_client.get(parameters.group_name, parameters.service_plan_name)
+            self.log.info("Found app insights %s in %s resource group" %
+                          (parameters.service_plan_name, parameters.group_name))
+        except Exception:
+            self.log.info("Deploying app insights %s in %s resource group" %
+                          (parameters.service_plan_name, parameters.group_name))
+            params = {
+                'location': parameters.app_insights_location,
+                'application_type': parameters.webapp_name,
+                'request_source': 'IbizaWebAppExtensionCreate',
+                'kind': 'web'
+            }
+            ai_client.components.create_or_update(parameters.group_name,
+                                                  parameters.service_plan_name,
+                                                  params)
+
+        # Deploy App Service Plan
+        web_client = session.client('azure.mgmt.web.WebSiteManagementClient')
+        service_plan = \
```

Comment:
This is the same as above. I think if we don't find the resource group then we don't need to keep checking if these other resources exist. If the rg doesn't exist, we can already move on and start provisioning the resources.